blockchain/indexers/indexsubscriber.go

```diff
@@ -348,43 +350,85 @@ func (s *IndexSubscriber) CatchUp(ctx context.Context, db database.DB, queryer C
 	return nil
 }
 
-// Run relays index notifications to subscribed indexes.
+// handleSyncSubscribers updates index sync subscribers when a subscribed
+// indexer is fully synced.
 //
 // This should be run as a goroutine.
-func (s *IndexSubscriber) Run(ctx context.Context) {
+func (s *IndexSubscriber) handleSyncSubscribers(ctx context.Context) {
+	ticker := time.NewTicker(time.Millisecond * 500)
+
 	for {
 		select {
-		case ntfn := <-s.c:
-			// Relay the index update to subscribed indexes.
+		case <-ctx.Done():
+			ticker.Stop()
+			s.wg.Done()
+			return
+
+		case <-ticker.C:
+			s.mtx.Lock()
 			for _, sub := range s.subscriptions {
-				err := updateIndex(ctx, sub.idx, &ntfn)
+				err := maybeNotifySubscribers(ctx, sub.idx)
 				if err != nil {
-					log.Error(err)
+					s.mtx.Unlock()
+					log.Error("unable to notify sync subscribers: %v", err)
 					s.cancel()
-					break
 				}
 			}
+			s.mtx.Unlock()
+		}
+	}
+}
 
-			if ntfn.Done != nil {
-				close(ntfn.Done)
-			}
-
+// handleIndexUpdates relays updates to subscribed indexers.
+func (s *IndexSubscriber) handleIndexUpdates(ctx context.Context) {
+	for {
+		select {
 		case <-ctx.Done():
-			log.Infof("Index subscriber shutting down")
-
 			close(s.quit)
 
 			// Stop all updates to subscribed indexes and terminate their
 			// processes.
+			s.mtx.Lock()
 			for _, sub := range s.subscriptions {
 				err := sub.stop()
 				if err != nil {
+					s.mtx.Unlock()
```

Comment:
It's unlocking here without a return and unlocking again later which will cause a panic.